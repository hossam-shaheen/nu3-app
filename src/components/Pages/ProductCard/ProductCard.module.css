.product-card-container {
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 100vh;
}

.product-card {
  display: flex;
  justify-content: space-between;
  background-color: #fff;
  width: 100%;
  padding: 20px 40px;
  position: relative;
}

.product-gallery {
  width: 40%;
  display: flex;
  flex-direction: column;
  align-items: center;

}

.product-info {
  width: 50%;
}

.product-attribute-qr {
  display: flex;
  justify-content: space-between;
}

.back {
  position: absolute;
  top: -20px;
  left: -10px;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background-color: var(--secondary);
  font-size: 18px;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;

}