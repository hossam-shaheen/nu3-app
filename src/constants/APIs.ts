export const SEARCH_BASE_URL = 'http://localhost:3000/data/search.json';
export const PRODUCT_BASE_URL = 'http://localhost:3000/data/product.json';