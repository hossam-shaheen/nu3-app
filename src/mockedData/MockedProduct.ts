export const mockedProduct = {
  product: {
    id: 4365311672401,
    title: "nu3 Bio Acerola",
    body_html:
      "nu3 Bio Acerola-Pulver: 100 g ✓ Natürliche Vitamin C-Quelle ✓ Ohne Zusatzstoffe wie Maltodextrin ✓ Schonend gefriergetrocknet. Jetzt im nu3 Shop bestellen!",
    vendor: "nu3",
    url: "https://www.nu3.de/products/nu3-bio-acerola-pulver",

    variants: [
      {
        id: 32511433474129,
        title: "150 g",
        price: "18.99",
        option1: "150 g",
      },
      {
        id: 32788214087761,
        title: "3 x 150 g",
        price: "51.99",
        option1: "3 x 150 g",
      },
    ],
    options: [
      {
        id: 5668092411985,
        values: ["150 g", "3 x 150 g"],
      },
    ],
    images: [
      {
        id: 15491638657105,
        src: "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085",
      },
      {
        id: 16206942535761,
        src: "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/a_252F2_252F4_252Ff_252Fa24f85e82f207a80c119cb64b0fa99d3ac4def70_21452.jpg?v=1630500198",
      },
    ],
    image: {
      id: 15491638657105,
      src: "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085",
    },
  },
};
