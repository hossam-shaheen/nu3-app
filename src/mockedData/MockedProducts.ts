export const mockedProducts = {
  items: [
    {
      price: 50.99,
      image:
        "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/8_252F1_252Fc_252Fc_252F81ccf47420c1f3a11717ff5ad7f93e3e13214556_21446_large.jpg?v=1620826303",
      vendor: "nu3",
      title: "nu3 Premium Skin & Hair Beauty",
      id: "4725777367121",
      link: "/products/nu3-premium-skin-and-hair-beauty",
    },
  ],
};
