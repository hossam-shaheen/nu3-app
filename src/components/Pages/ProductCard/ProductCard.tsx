import { FunctionComponent, useCallback, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import ProductAttribute from "src/components/Components/ProductCard/ProductAttribute/ProductAttribute";
import ProductGallery from "src/components/Components/ProductCard/ProductGallery/ProductGallery";
import ProductVariants from "src/components/Components/ProductCard/ProductVariants/ProductVariants";
import Error from "src/components/Components/Shared/Error/Error";
import Loader from "src/components/Components/Shared/Loader/Loader";
import QrCode from "src/components/Components/Shared/QrCode/QrCode";
import {
  ErrorType,
  ProductAttributeProps,
  ProductInfoType,
  ProductVariantsProps,
} from "src/interfaces/interfaces";
import { PRODUCT_BASE_URL } from "../../../constants/APIs";
import classes from "./ProductCard.module.css";

export const ProductCard: FunctionComponent = (): JSX.Element => {
  const [productInfo, setProductInfo] = useState<ProductInfoType | null>(null);
  const [error, setError] = useState<ErrorType | null>(null);
  const navigate = useNavigate();

  const getCurrentProduct = useCallback(async () => {
    try {
      const response = await fetch(`${PRODUCT_BASE_URL}`);
      const data = await response.json();
      setProductInfo(data.product);
    } catch (error) {
      setError({
        icon: "fas fa-exclamation-triangle",
        message: "Failed to fetch data",
      });
    }
  }, []);

  useEffect(() => {
    getCurrentProduct();
  }, [getCurrentProduct]);

  const productAttributes: ProductAttributeProps | null = productInfo && {
    productImage: productInfo?.image,
    productImages: productInfo?.images,
    ProductTitle: productInfo?.title,
    productBodyHtml: productInfo?.body_html,
    productVendor: productInfo?.vendor,
  };

  const productVariants: ProductVariantsProps | null = productInfo && {
    variants: productInfo?.variants,
    options: productInfo?.options,
  };
  
  const back = () => {
    navigate(-1);
  };

  return (
    <div className={classes["product-card-container"]}>
      <div className={classes["product-card"]}>
        {!error && productInfo && (
          <>
            <div className={classes["product-gallery"]}>
              <ProductGallery
                mainImage={productAttributes?.productImage}
                thumbnails={productAttributes?.productImages}
              />
            </div>
            <div className={classes["product-info"]}>
              <div className={classes["product-attribute-qr"]}>
                <ProductAttribute productAttributes={productAttributes} />
                <QrCode url={`${productInfo?.url}`} />
              </div>
              <ProductVariants productVariants={productVariants} />
              <button>
                <i className="fas fa-shopping-cart"></i> Add To Cart
              </button>
            </div>
          </>
        )}

        {error && !productInfo && (
          <Error errorClass="error-message" error={error} />
        )}
        {!error && !productInfo && <Loader />}

        <button className={classes["back"]} onClick={back} data-testid="back">
          <i className="fas fa-chevron-left"></i>
        </button>
      </div>
    </div>
  );
};
export default ProductCard;
