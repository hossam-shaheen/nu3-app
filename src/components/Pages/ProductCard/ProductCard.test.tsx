import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import { setupServer } from "msw/node";
import { rest } from "msw";
import ProductCard from "./ProductCard";
import { mockedProduct } from "src/mockedData/MockedProduct";
import { PRODUCT_BASE_URL } from "src/constants/APIs";
import { BrowserRouter } from "react-router-dom";

const server = setupServer(
  rest.get(`${PRODUCT_BASE_URL}`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(mockedProduct));
  })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("ProductCard product data should be displayed : Renders", async () => {
  render(
    <BrowserRouter>
      <ProductCard />
    </BrowserRouter>
  );

  const productTitle = await screen.findByRole("heading");
  expect(productTitle).toBeInTheDocument();

  const productVendor = await screen.findByText("nu3");
  expect(productVendor).toBeInTheDocument();

  const productQrCode = await screen.findByTestId("QR-Code-SVG");
  expect(productQrCode).toBeInTheDocument();

  const productHtmlList = screen.getAllByRole("listitem");
  expect(productHtmlList.length).toBeGreaterThan(0);

  const productImages = screen.queryAllByRole("img");
  expect(productImages.length).toBeGreaterThan(0);

  const productButton = screen.getByText("Add to cart", { exact: false });
  expect(productButton).toBeInTheDocument();

  const backButton = screen.getByTestId("back");
  expect(backButton).toBeInTheDocument();
});

test("ProductCard : Snapshot", () => {
  const component = renderer.create(
    <BrowserRouter>
      <ProductCard />
    </BrowserRouter>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
