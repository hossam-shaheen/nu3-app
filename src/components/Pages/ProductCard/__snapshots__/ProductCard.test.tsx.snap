// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard : Snapshot 1`] = `
<div
  className="product-card-container"
>
  <div
    className="product-card"
  >
    <div
      className="loading"
    >
      Loading ...
    </div>
    <button
      className="back"
      data-testid="back"
      onClick={[Function]}
    >
      <i
        className="fas fa-chevron-left"
      />
    </button>
  </div>
</div>
`;
